clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,38 +603,166 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
     if (concurrencyKey == DEFAULT_KEY) {
       stateService.getStateRepositoryManager().syncMessageFor(storeIdentifier).forEach(syncChannel::synchronizeToPassive);
     } else {
+      boolean interrupted = false;
+      BlockingQueue<DataSyncMessageHandler> messageQ = new SynchronousQueue<>();
       int segmentId = concurrencyKey - DEFAULT_KEY - 1;
-      Long dataSizeThreshold = Long.getLong(SYNC_DATA_SIZE_PROP, DEFAULT_SYNC_DATA_SIZE_THRESHOLD);
-      AtomicLong size = new AtomicLong(0);
+      Thread thisThread = Thread.currentThread();
+      CompletableFuture<Void> asyncGet = CompletableFuture.runAsync(
+        () -> doGetsForSync(segmentId, messageQ, syncChannel, thisThread), SYNC_GETS_EXECUTOR);
+      try {
+        try {
+          while (messageQ.take().execute()) ;
```

Comment:
I was hoping that moving to this model would allow you to eliminate most of the interrupt logic. If the async get thread hits an issue it can just push a task to the queue that throws the same exception, thereby causing the termination of the writer thread.

Reply:
@chrisdennis , yes I did get your intention. But since it is a synchronousQueue (where the put succeeds only if other side is waiting on a take - the other side could be in the synchronizeToPassive), it becomes hard to do that. Also to add to the complexity, I have to catch the interrupt exception again inside throwable. 

May have to use some other queue other than a synchronous queue to cleanly do that, which brings other complexities.

Thought it was simpler to still use synchronous queue (as it is the simplest way to do a double buffering mechanism). The error handling works as the test I added suggests. 

Besides any error here is highly unlikely and results in a RuntimeException which results in a server shutdown.

Again suggestions welcome, as I might have missed something.